rocAL/source/readers/image/numpy_data_reader.cpp

```diff
@@ -0,0 +1,587 @@
+        _last_id.erase(0, last_slash_idx + 1);
+    }
+
+    auto ret = get_header_from_cache(file_path, _curr_file_header);
+    if (!ret) {
+        parse_header(_curr_file_header, file_path);
+        update_header_cache(file_path, _curr_file_header);
+    } else {
+        _current_fPtr = std::fopen(file_path.c_str(), "rb");
+        if (_current_fPtr == nullptr)
+            THROW("Could not open file " + file_path + ": " + std::strerror(errno));
+    }
+    fseek(_current_fPtr, 0, SEEK_SET);  // Take the file pointer back to the start
+
+    return _curr_file_header.nbytes();
+}
+
+bool NumpyDataReader::get_header_from_cache(const std::string& file_name, NumpyHeaderData& header) {
+    std::unique_lock<std::mutex> cache_lock(_cache_mutex);
+    auto it = _header_cache.find(file_name);
+    if (it == _header_cache.end()) {
+        return false;
+    } else {
+        header = it->second;
+        return true;
+    }
+}
+
+void NumpyDataReader::update_header_cache(const std::string& file_name, const NumpyHeaderData& value) {
+    std::unique_lock<std::mutex> cache_lock(_cache_mutex);
+    _header_cache[file_name] = value;
+}
+
+const RocalTensorDataType NumpyDataReader::get_dtype(const std::string& format) {
+    if (format == "u1") return RocalTensorDataType::UINT8;
+    if (format == "u2") THROW("uint16_t dtype not supported in rocAL");
+    if (format == "u4") return RocalTensorDataType::UINT32;
+    if (format == "u8") THROW("uint64_t dtype not supported in rocAL");
+    if (format == "i1") return RocalTensorDataType::INT8;
+    if (format == "i2") return RocalTensorDataType::INT16;
+    if (format == "i4") return RocalTensorDataType::INT32;
+    if (format == "i8") THROW("int64_t dtype not supported in rocAL");
+    if (format == "f2")
+#if defined(AMD_FP16_SUPPORT)
+        return RocalTensorDataType::FP16;
+#else
+        THROW("FLOAT16 type tensor not supported")
+#endif
+    if (format == "f4") return RocalTensorDataType::FP32;
+    if (format == "f8") THROW("double dtype not supported in rocAL");
+    THROW("Unknown Numpy dtype string");
+}
+
+inline void NumpyDataReader::skip_spaces(const char*& ptr) {
+    while (::isspace(*ptr))
+        ptr++;
+}
+
+template <size_t N>
+void NumpyDataReader::skip_char(const char*& ptr, const char (&what)[N]) {
+    if (strncmp(ptr, what, N - 1))
+        THROW("Found wrong symbol during parsing");
+    ptr += N - 1;
+}
+
+template <size_t N>
+bool NumpyDataReader::try_skip_char(const char*& ptr, const char (&what)[N]) {
+    if (!strncmp(ptr, what, N - 1)) {
+        ptr += N - 1;
+        return true;
+    } else {
+        return false;
+    }
+}
+
+template <size_t N>
+void NumpyDataReader::skip_field(const char*& ptr, const char (&name)[N]) {
+    skip_spaces(ptr);
+    skip_char(ptr, "'");
+    skip_char(ptr, name);
+    skip_char(ptr, "'");
+    skip_spaces(ptr);
+    skip_char(ptr, ":");
+    skip_spaces(ptr);
+}
+
+template <typename T>
+T NumpyDataReader::parse_int(const char*& ptr) {
```

Comment:
Do we have only one function call for this? If so can we add as inline or add it in the respective function itself?

Reply:
I checked online and found that templates are implicitly inlined